### PR TITLE
Memory footprint

### PR DESCRIPTION
**Pull request by breznak**
- Connections use `unsigned char` for SynapseIdx, SegmentIdx
  - this limits synapses per segment, segments per cells to 255; which sounds reasonable
- SP remove tieBreaker vector
  - and fixing resulting exact output checks

For #360

### CODE REVIEW

File: src/examples/hotgym/HelloSPTP.cpp [hunk 1]
```diff
@@ -166,29 +166,29 @@ Real64 BenchmarkHotgym::run(UInt EPOCHS, bool useSPlocal, bool useSPglobal, bool
...
+        0, 4, 13, 21, 24, 30, 32, 37, 40, 46, 47, 48, 50, 51, 64, 68, 79, 81, 89, 97, 99, 114, 120, 135, 136, 140, 141, 143, 144, 147, 151, 155, 161, 162, 164, 165, 169, 172, 174, 179, 181, 192, 201, 204, 205, 210, 213, 226, 227, 237, 242, 247, 249, 254, 255, 262, 268, 271, 282, 283, 295, 302, 306, 307, 317, 330, 349, 353, 366, 380, 383, 393, 404, 409, 410, 420, 422, 441,446, 447, 456, 458, 464, 468, 476, 497, 499, 512, 521, 528, 531, 534, 538, 539, 541, 545, 550, 557, 562, 565, 575, 581, 589, 592, 599, 613, 617, 622, 647, 652, 686, 687, 691, 699, 704, 710, 713, 716, 722, 729, 736, 740, 747, 749, 753, 754, 758, 766, 778, 790, 791, 797, 800, 808, 809, 812, 815, 826, 828, 830, 837, 852, 853, 856, 863, 864, 873, 878, 882, 885, 893, 894, 895, 905, 906, 914, 915, 920, 924, 927, 937, 939, 944, 947, 951, 954, 956, 967, 968, 969, 973, 975, 976, 979, 981, 991, 998
```
Comment (breznak) [inline src/examples/hotgym/HelloSPTP.cpp]: encoder changed, because I had to change seed; otherwise TM gave no output in this configuration. 
Proper fix queued.

File: src/htm/algorithms/Connections.cpp [hunk 2]
```diff
@@ -199,7 +199,7 @@ void Connections::destroySegment(Segment segment) {
   const auto segmentOnCell =
       std::lower_bound(cellData.segments.begin(), cellData.segments.end(),
                        segment, [&](Segment a, Segment b) {
-                         return segmentOrdinals_[a] < segmentOrdinals_[b];
+                         return segmentOrdinals_[a] < segmentOrdinals_[b]; //TODO will this be slow if ordinals moved to SegmentData?
```
Comment (breznak) [inline src/htm/algorithms/Connections.cpp]: ? My idea is to get rid of the *Ordinals_ arrays in Conn, and keep that number in SegmentData

File: src/htm/algorithms/Connections.hpp [hunk 3]
```diff
@@ -36,9 +35,10 @@
 namespace htm {
 
 //TODO instead of typedefs, use templates for proper type-checking?
+//TODO profile to use better (smaller?) types
 using CellIdx   = htm::ElemSparse; // CellIdx must match with ElemSparse, defined in Sdr.hpp
-using SegmentIdx= UInt16; /** Index of segment in cell. */
-using SynapseIdx= UInt16; /** Index of synapse in segment. */ //TODO profile to use better (smaller?) types
+using SegmentIdx= unsigned char; /** Index of segment in cell. */
+using SynapseIdx= unsigned char; /** Index of synapse in segment. */
```
Comment (breznak) [inline src/htm/algorithms/Connections.hpp]: the new type, only 1B!. If should overflow we have check to inform

File: src/htm/algorithms/SpatialPooler.cpp [hunk 4]
```diff
@@ -363,7 +363,7 @@ void SpatialPooler::getConnectedSynapses(UInt column,
 void SpatialPooler::getConnectedCounts(UInt connectedCounts[]) const {
   for(UInt seg = 0; seg < numColumns_; seg++) {
     const auto &segment = connections_.dataForSegment( seg );
-    connectedCounts[ seg ] = segment.numConnected;
+    connectedCounts[ seg ] = segment.numConnected; //TODO numConnected only used here, rm from SegmentData and compute for each segment.synapses?
```
Comment (breznak) [inline src/htm/algorithms/SpatialPooler.cpp]: plan to test if this should be moved from SegData, and computed only as a lambda here. 
This is the only place where this is used, but updated each updateSynapsePermanence

File: src/htm/algorithms/SpatialPooler.cpp [hunk 5]
```diff
@@ -854,11 +849,6 @@ void SpatialPooler::inhibitColumnsGlobal_(const vector<Real> &overlaps,
   NTA_ASSERT(!overlaps.empty());
   NTA_ASSERT(density > 0.0f && density <= 1.0f);
 
-  // Add a tiebreaker to the overlaps so that the output is deterministic.
-  vector<Real> overlaps_(overlaps.begin(), overlaps.end());
-  for(UInt i = 0; i < numColumns_; i++)
-    overlaps_[i] += tieBreaker_[i];
```
Comment (breznak) [inline src/htm/algorithms/SpatialPooler.cpp]: removed tieBreakers. 
+ saves memory
+ makes inhibition faster

**Comment by breznak** [comment]
> [ RUN      ] TMRegionTest.testLinking
In VectorFileSensor TestOutputDir/TMRegionTestInput.csv
Reading CSV file
Read 10 vectors
unknown file: error: SEH exception with code 0xc0000005 thrown in the test body.

some weird error on Windows CI, restarting in hope it's just random

**Comment by breznak** [comment]
@dkeeney if you have time, could you please test this on a local Windows machine, please? THe PR keeps throwing a weird err on Windows CI or crashing the CI build, while the PR is relatively easy and I don't see a reason why this should happen.

**Comment by dkeeney** [comment]
> could you please test this on a local Windows machine, please?

Sure, I will try it out.

**Comment by dkeeney** [comment]
Ok, here is the problem:   in `TMRegionTest.cpp` line 292
```
  std::vector<Byte> expected3out = VectorHelpers::sparseToBinary<Byte>(
            {
	     0, 1, 2, 3, 4, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 30, 
	     31, 32, 33, 34, 35, 36, 37, 38, 39, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 70,
	     71, 72, 73, 74, 85, 86, 87, 88, 890, 1, 2, 3, 4, 10, 11, 12, 13, 14, 15, 16, 17,
	     18, 19, 20, 21, 22, 23, 24, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 50, 51, 52, 53, 
	     54, 55, 56, 57, 58, 59, 70, 71, 72, 73, 74, 85, 86, 87, 88, 89 
	     }, (UInt32)r3OutputArray.getCount());
```
The r3OutputArray size is 100.  
There is a typo in the middle of the initialization string.  
The value `890` should be `89`.

This sparseToBinary( ) function probably should have checked for index-out-of-range at least in debug mode.

I will let you make the change.

**Comment by breznak** [comment]
> The value 890 should be 89.

wow, you have an eye of a bug hunter :) Thank you!

Is the SEH kind of segfault error on windows? I was suprised I didn't find anything like that. 

Also strange it passes (?) of Unixes. 

I had a PR getting rid of SparseToBinary in favour of SDR, should revisit it once things clear up

File: src/examples/hotgym/HelloSPTP.cpp [hunk 6]
```diff
@@ -64,7 +64,7 @@ Real64 BenchmarkHotgym::run(UInt EPOCHS, bool useSPlocal, bool useSPglobal, bool
   SpatialPooler  spLocal(enc.dimensions, vector<UInt>{COLS}); // Spatial pooler with local inh
   spGlobal.setGlobalInhibition(true);
   spLocal.setGlobalInhibition(false);
-  Random rnd(1); //uses fixed seed for deterministic output checks
+  Random rnd(42); //uses fixed seed for deterministic output checks
```
Comment (dkeeney) [inline src/examples/hotgym/HelloSPTP.cpp]: Ah, back to 42 ... "the meaning of life" 👍

File: src/htm/algorithms/SpatialPooler.hpp [hunk 7]
```diff
@@ -287,7 +287,6 @@ class SpatialPooler : public Serializable
     ar(CEREAL_NVP(overlapDutyCycles_));
     ar(CEREAL_NVP(activeDutyCycles_));
     ar(CEREAL_NVP(minOverlapDutyCycles_));
-    ar(CEREAL_NVP(tieBreaker_));
```
Comment (dkeeney) [inline src/htm/algorithms/SpatialPooler.hpp]: Since tieBreaker is deterministic it does not need to be saved so this is correct.  But are you sure it is regenerated after the restore?  If it should be different then answers will eventually drift off of the correct set on subsequent iterations.

File: src/htm/utils/VectorHelpers.hpp [hunk 8]
```diff
@@ -26,6 +26,7 @@ class VectorHelpers
   {
     std::vector<T> binary(width);
     for (auto sparseIdx: sparseVector) {
+      NTA_ASSERT(sparseIdx < binary.size()) << "attemping to insert out of bounds element! " << sparseIdx;
```
Comment (dkeeney) [inline src/htm/utils/VectorHelpers.hpp]: 👍

File: src/test/unit/regions/TMRegionTest.cpp [hunk 9]
```diff
@@ -287,12 +289,13 @@ TEST(TMRegionTest, testLinking) {
...
-             25u, 26u, 27u, 28u, 29u, 30u, 31u, 32u, 33u, 34u, 40u, 41u, 42u, 43u,
-             44u, 55u, 56u, 57u, 58u, 59u, 65u, 66u, 67u, 68u, 69u, 85u, 86u, 87u,
-             88u, 89u, 95u, 96u, 97u, 98u, 99u }, (UInt32)r3OutputArray.getCount());
-  EXPECT_TRUE(r3OutputArray == expected3out);
+  std::vector<Byte> expected3out = VectorHelpers::sparseToBinary<Byte>( //TODO replace with SDR 
+            {
+	     0, 1, 2, 3, 4, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 30, 
+	     31, 32, 33, 34, 35, 36, 37, 38, 39, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 70,
+	     71, 72, 73, 74, 85, 86, 87, 88, 89
+	     }, (UInt32)r3OutputArray.getCount());
```
Comment (dkeeney) [inline src/test/unit/regions/TMRegionTest.cpp]: Good, you corrected the paste error :)

File: src/htm/algorithms/SpatialPooler.hpp (hunk 7 above)
Comment (breznak) [reply src/htm/algorithms/SpatialPooler.hpp]: no, tie breakers are totally removed and not used, as they did not have 100% the effect we expected. And you've obsoleted them in the long-dragging Deterministic builds PR  :+1: where you figured we need to add the `a > b` to the segment ordering algorithm. 

This is because tie breakers sometimes work and break ties:
{1,1} + {0, 0.01} -> L < R

but sometimes they just break (we apply tie breaks flat out to all, for performance reasons)
{1, 0.99} + {0, 0.01} -> L ? R broken

TL;DR this PR removes tieBreakers altogether because those are not needed (we do deterministic sort)

Comment (breznak) [reply src/htm/algorithms/SpatialPooler.hpp]: see the removal https://github.com/htm-community/htm.core/pull/534#discussion_r299778927

Comment (dkeeney) [reply src/htm/algorithms/SpatialPooler.hpp]: Oh, somehow I missed the fact that they were removed.  I like that you did that.

**Comment by ctrl-z-9000-times** [comment]
I would really like to see some measurements of the run-time & memory usage of an HTM before and after these changes.  Tests using real world data would be even better.  Optimizing code without measuring the results is haphazard.

**Comment by dkeeney** [comment]
> Is the SEH kind of segfault error on windows?

There are two exception facilities on windows, SEH is the one that handles most hardware faults.  We use the other one EHsc ...the one that handles only std::exceptions per the standard.  C++ cannot use both facilities at the same time so hardware faults (like index out of range and divide by 0) will cause crashes.  

I suspect the reason Linux did not see this error is that the offset just happened to fall within some other accessible data space...would have caused a really hard to find bug if it hit something important.

**Comment by breznak** [comment]
> I would really like to see some measurements of the run-time & memory usage of an HTM before and after these changes. Tests using real world data would be even better.  

- tie breakers are now useless and should just go. Saves a bit on computation and mem. 

- 255 synapses per seg, segments per cell: 
 well, my ideal was playing with very large models. Also I thought less storage -> more fit in cpu cache. 
I have never seen a real world model here that would exceed the 255 limit, so I thought why not. 
Real world impact is quite minor (2B per segment). 

> Figure 6 illustrates the performance of a network of HTM neurons implementing a high-order sequence memory. The network used in Figure 6 consists of 2048 mini-columns with 32 neurons per mini-column. Each neuron has 128 basal dendritic segments, and each dendritic segment has up to 40 actual synapses

https://www.frontiersin.org/articles/10.3389/fncir.2016.00023/full 

To sum up, I think we can back up the "to char" change. If we never need it anyway, we could use the smaller version, if there is a possible scenario where 1000s synapses per segment would be useful, let's return to uint16. 

SynapseData is a different beast, which is much more populous,  #360

**Comment by ctrl-z-9000-times** [comment]
> if there is a possible scenario where 1000s synapses per segment would be useful, let's return to uint16.

Segments should not have 1000s of synapses, but they could have 1000s of _potential_ synapses.

This is only an issue for the spatial pooler which makes synapses for the whole potential pool.  The temporal memory makes the synapses as needed.

File: src/htm/algorithms/SpatialPooler.cpp (hunk 4 above)
Comment (ctrl-z-9000-times) [reply src/htm/algorithms/SpatialPooler.cpp]: I'd like to keep `connectedCounts` computed and on hand
- The synapse competition uses it
- The (Kropff & Treves 2008) model uses it
- It's fast to compute, its a small amount of book keeping that gets done when a synapse changes between connected/disconnected

Maybe these applications can be optimized to not need this data, or maybe there is a different way to optimize this, but for now I'd rather not remove this.

Comment (breznak) [reply src/htm/algorithms/SpatialPooler.cpp]: > The synapse competition uses it
The (Kropff & Treves 2008) model uses it

sure, let's keep it if you have coming users for that!

File: src/htm/algorithms/Connections.hpp (hunk 3 above)
Comment (breznak) [reply src/htm/algorithms/Connections.hpp]: thanks for calling me out on this! Reverted back to UInt16, as it's more usable for many potential synapses. Also the UInt16 actually runs faster.

**Comment by breznak** [comment]
Reverted back to UInt16. This is now only the tieBreaker removal PR. Please rereview

File: src/htm/algorithms/SpatialPooler.cpp [hunk 10]
```diff
@@ -429,12 +429,7 @@ void SpatialPooler::initialize(
   iterationNum_ = 0u;
   iterationLearnNum_ = 0u;
 
-  tieBreaker_.resize(numColumns_);
-  for (Size i = 0; i < numColumns_; i++) {
-    tieBreaker_[i] = (Real)(0.01 * rng_.getReal64());
-  }
-
-  overlapDutyCycles_.assign(numColumns_, 0);
+  overlapDutyCycles_.assign(numColumns_, 0); //TODO make all these sparse or rm to reduce footprint
```
Comment (ctrl-z-9000-times) [inline src/htm/algorithms/SpatialPooler.cpp]: I think we can get rid of `overlapDutyCycles_` altogether.  IIRC it is a boosting method?  We have better/stronger boosting methods.  I don't think that this even ever kicks in?  But that needs to be checked before removing it, an issue for a different PR

Comment (breznak) [reply src/htm/algorithms/SpatialPooler.cpp]: I think it's used for something...bumpupWeakSegments or so.. I have waiting another PR that puts all of boosting code into 2 methods: one for boosting overlaps, other for weak columns (and shows the "weak columns" is not really useful, on the limited mnist, hotgym tests) But yes, another PR

**Comment by breznak** [comment]
Is this good to go?

**Comment by breznak** [comment]
Thank you both for good reviews and help with the PR!
